### PR TITLE
Add processor status checker

### PR DESCRIPTION
**Pull request by rtso**
### Description
Add a post processing metric to track Hasura's latest updated time for each network

### Test Plan
`cargo run --release -- -c config.yaml` 

Look at the metric output from http://0.0.0.0:8088/metrics
Screenshot 2023-08-22 at 11 44 45 AM

### CODE REVIEW

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/metrics.rs [hunk 1]
```diff
@@ -35,3 +35,13 @@ pub static TASK_FAILURE_COUNT: Lazy<IntCounterVec> = Lazy::new(|| {
     )
     .unwrap()
 });
+
+// Processor last update time latency to current time in seconds.
+pub static PROCESSOR_LAST_UPDATED_TIME_LATENCY_IN_SECS: Lazy<GaugeVec> = Lazy::new(|| {
+    register_gauge_vec!(
+        "indexer_grpc_post_processor_processor_last_updated_time_latency_in_secs",
```
Comment (just-in-chang) [inline ecosystem/indexer-grpc/indexer-grpc-post-processor/src/metrics.rs:42]: nit: 2 `processor`s

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        Err(e) => {
+            tracing::error!(e = ?e, "Failed to get processor status response");
+            panic!();
+        },
+    };
+
+    for processor_status in processor_status_response.processor_status {
+        let last_updated_time = NaiveDateTime::parse_from_str(
+            processor_status.last_updated.as_str(),
+            // Parse "2023-08-12T19:49:42.580696"
```
Comment (just-in-chang) [inline ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs:80]: can prob take this comment out too

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    pub fn new(hasura_rest_api_endpoints: Vec<String>) -> Self {
+        Self {
+            hasura_rest_api_endpoints,
+        }
+    }
+
+    pub async fn run(&self) -> Result<()> {
+        loop {
+            let mut handles = Vec::new();
+            // Connect to each hasura endpoing in config
```
Comment (just-in-chang) [inline ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs:48]: endpoing 🙂

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+use std::{
+    collections::BTreeMap,
+    ops::Div,
+    sync::Arc,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use tracing::info;
+
+const PROCESSOR_STATUS_CHECKER_WAIT_TIME_IN_SECS: u64 = 10;
+const GET_LATEST_PROCESSOR_STATUS: &str = "/api/rest/get_lastest_processor_status";
```
Comment (just-in-chang) [inline ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs:19]: can prob move these to `ProcessorStatusCheckerConfig`? or maybe just the first wait time

**Review by just-in-chang — APPROVED** [review]

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/metrics.rs (hunk 1 above)
Comment (rtso) [reply ecosystem/indexer-grpc/indexer-grpc-post-processor/src/metrics.rs:42]: lol "indexer_grpc_post_processor" is the namespace and "processor_last_updated_time_latency_in_secs" is the name of the metric. 

I'll rename it to "indexer_processors_processor_last_updated_time_latency_in_secs"

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/config.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+            - http://fullnode.2.address/v1
+            - http://fullnode.3.address/v1
+        indexer_grpc_address: IP:PORT
+        indexer_grpc_auth_token: AUTH_TOKEN
+    processor_status_checker_config:
+        # "{endpoint}/api/rest/get_lastest_processor_status" where endpoint is one of:
+        #     - https://indexer.mainnet.aptoslabs.com,
+        #     - https://indexer-testnet.staging.gcp.aptosdev.com
+        #     - https://indexer-devnet.staging.gcp.aptosdev.com
+        hasura_rest_api_endpoint: "https://indexer.mainnet.aptoslabs.com/api/rest/get_lastest_processor_status"
```
Comment (jillxuu) [inline ecosystem/indexer-grpc/indexer-grpc-post-processor/config.yaml:21]: nit: can we have a empty line at the end of the file?

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/main.rs [hunk 6]
```diff
@@ -88,6 +95,23 @@ impl RunnableConfig for IndexerGrpcPostProcessorConfig {
             }));
         }
 
+        if let Some(config) = &self.processor_status_checker_config {
+            tasks.push(tokio::spawn({
+                let config = config.clone();
+                async move {
+                    let checker =
+                        ProcessorStatusChecker::new(config.hasura_rest_api_endpoint.clone());
+                    info!("Starting ProcessorStatusChecker");
```
Comment (jillxuu) [inline ecosystem/indexer-grpc/indexer-grpc-post-processor/src/main.rs:104]: is it ok to expose `hasura_rest_api_endpoint` in the `info`?

**Review by larry-aptos — APPROVED** [review]

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs [hunk 7]
```diff
@@ -0,0 +1,85 @@
+// Copyright © Aptos Foundation
+
```
Comment (jillxuu) [inline ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs:2]:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    pub hasura_rest_api_endpoint: String,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct ProcessorStatusResponse {
+    processor_status: Vec<ProcessorStatus>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+
```
Comment (jillxuu) [inline ecosystem/indexer-grpc/indexer-grpc-post-processor/src/processor_status_checker.rs:22]:
```suggestion
```

**Review by jillxuu — APPROVED** [review]

File: ecosystem/indexer-grpc/indexer-grpc-post-processor/src/main.rs (hunk 6 above)
Comment (rtso) [reply ecosystem/indexer-grpc/indexer-grpc-post-processor/src/main.rs:104]: yes I think it's fine since it's a publicly available endpoint.

**Review by bowenyang007 — CHANGES_REQUESTED** [review]
Let's talk more about this. The grpc post processor doesn't seem to be the right place for this. Also there's already a cloudrun doing similar things.

**Comment by rtso** [comment]
Spoke to @geekflyer and it seems like the existing cloud run doesn't quite fit this use case. We plan to setup a new cloud run instance for post processing metrics. cc @bowenyang007

**Comment by rtso** [comment]
I'm going to move this over to a crate in the aptos-indexer-processors repo

**Pull request closed without merging**